### PR TITLE
[cloud-provider] fix test failed because of flag redefined

### LINKED ISSUE #101927: [Failing test] k8s.io/legacy-cloud-providers/vsphere TestSecretUpdated
##### Error text:
```
=== RUN   TestSecretUpdated
/tmp/go-build2183127444/b5733/vsphere.test flag redefined: log_dir
--- FAIL: TestSecretUpdated (0.05s)
panic: /tmp/go-build2183127444/b5733/vsphere.test flag redefined: log_dir [recovered]
	panic: /tmp/go-build2183127444/b5733/vsphere.test flag redefined: log_dir

goroutine 1848 [running]:
testing.tRunner.func1.2(0x2a1d100, 0xc000a30250)
	/usr/local/go/src/testing/testing.go:1143 +0x49f
testing.tRunner.func1(0xc000560300)
	/usr/local/go/src/testing/testing.go:1146 +0x695
panic(0x2a1d100, 0xc000a30250)
	/usr/local/go/src/runtime/panic.go:971 +0x499
flag.(*FlagSet).Var(0xc00012a180, 0x32a2d28, 0x449fcb0, 0x2eb0fd6, 0x7, 0x2f2c182, 0x2f)
	/usr/local/go/src/flag/flag.go:871 +0x637
flag.(*FlagSet).StringVar(...)
	/usr/local/go/src/flag/flag.go:760
k8s.io/kubernetes/vendor/k8s.io/klog/v2.InitFlags(0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:427 +0xfb
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/vsphere.TestSecretUpdated(0xc000560300)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:1209 +0x58a
testing.tRunner(0xc000560300, 0x2fca618)
	/usr/local/go/src/testing/testing.go:1193 +0x203
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1238 +0x5d8

```
#### Recent failures:
[5/11/2021, 9:25:38 AM ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1392153616399208448)
[5/11/2021, 3:25:18 AM ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1392062891410067456)
[5/10/2021, 9:25:56 PM ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1391972293101490176)
[5/10/2021, 3:25:43 PM ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1391881696697126912)
[5/10/2021, 9:25:30 AM ci-kubernetes-generate-make-test-count10](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-generate-make-test-count10/1391791097046372352)

/kind failing-test
/kind flake
/sig cloud-provider
/area provider/vmware

Flaked on https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101919/pull-kubernetes-unit/1392207748908716032

Is this the same issue as https://github.com/kubernetes/kubernetes/issues/98941 ?

### PR DESCRIPTION
**Pull request by jonyhy96**
#### What type of PR is this?

/kind bug
/kind failing-test
/kind flake

#### What this PR does / why we need it:

The test has continued to fail for months, better fix this.

#### Which issue(s) this PR fixes:

Fixes #101927

#### Special notes for your reviewer:

Already test this locally with 

before
```bash
$ go test -run ^TestSecretUpdated$ ./vsphere -count 10
W1122 11:02:32.160755    8326 vsphere.go:341] Global.User and Secret info provided. VCP will use secret to get credentials
W1122 11:02:32.160772    8326 vsphere.go:345] Global.Password and Secret info provided. VCP will use secret to get credentials
/var/folders/x0/3jhpxgj516nf3l_ns7jl09rr0000gn/T/go-build1721751768/b001/vsphere.test flag redefined: log_dir
--- FAIL: TestSecretUpdated (0.01s)
panic: /var/folders/x0/3jhpxgj516nf3l_ns7jl09rr0000gn/T/go-build1721751768/b001/vsphere.test flag redefined: log_dir [recovered]
        panic: /var/folders/x0/3jhpxgj516nf3l_ns7jl09rr0000gn/T/go-build1721751768/b001/vsphere.test flag redefined: log_dir

goroutine 75 [running]:
testing.tRunner.func1.2({0x26000c0, 0xc00069f390})
        /usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1212 +0x218
panic({0x26000c0, 0xc00069f390})
        /usr/local/Cellar/go/1.17.1/libexec/src/runtime/panic.go:1038 +0x215
flag.(*FlagSet).Var(0xc0000b4180, {0x2d7d848, 0x3f4f5f0}, {0x2a74599, 0x7}, {0x2af0911, 0x2f})
        /usr/local/Cellar/go/1.17.1/libexec/src/flag/flag.go:879 +0x2f4
flag.(*FlagSet).StringVar(...)
        /usr/local/Cellar/go/1.17.1/libexec/src/flag/flag.go:762
k8s.io/klog/v2.InitFlags(0x3f4f540)
        /Users/jony/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:429 +0x55
k8s.io/legacy-cloud-providers/vsphere.TestSecretUpdated(0xc000503a00)
        /Users/jony/work/kubernetes/staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:239 +0x405
testing.tRunner(0xc000503a00, 0x2b92f50)
        /usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1306 +0x35a
FAIL    k8s.io/legacy-cloud-providers/vsphere   0.760s
FAIL
```

after
```bash
$ go test -run ^TestSecretUpdated$ ./vsphere -count 10
ok      k8s.io/legacy-cloud-providers/vsphere   0.673s
```

### CODE REVIEW

**Comment by jonyhy96** [comment]
PTAL @cheftako

**Comment by aojea** [comment]
So, the problem seems that when we run the test multiple times `go test -count 10` , klog tries to set the log variable again
`vsphere.test flag redefined: log_dir` .

@pohly you were dealing with klog recently, can you advice on what will be the best way to deal with this problem so we can take a consistent solution?
@MikeSpreitzer this is another case of globals on unit test, this time klog instead of metrics :)

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go
Comment (pohly) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: `klog.InitFlags(nil)` can go into an `init` function. Then you don't need this `TestMain` at all.

Comment (jonyhy96) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: The `vsphere_test` is in package `vsphere`, if i use `init` here i'm afraid code can't define another `init` function for real logic init.

Comment (pohly) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: The `*_test.go` files are not getting compiled into the `vsphere` package so there's no risk of polutting the package.

You can also have more than one `init` function, even in the same file.

Comment (pohly) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: You just have to ensure that there is exactly one `init` function which calls `klog.InitFlags(nil)`.

Comment (jonyhy96) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: Thanks for explanation!

Comment (jonyhy96) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:211]: Done

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go
Comment (pohly) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:210]: You can remove `klog.Flush()`. There's nothing to flush yet.

**Comment by cheftako** [comment]
/lgtm
We should probably fix @pohly suggestion. 
I believe the failing test may just be #99877.

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Comment by jonyhy96** [comment]
PTAL @aojea @pohly @cheftako may need milestone 1.23 label

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go
Comment (jonyhy96) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go]: Done

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by aojea** [comment]
/lgtm
@cheftako we need your approval
/assign @liggitt @spiffxp 

do we want this in 1.23? seems we've missed test freeze

**Comment by liggitt** [comment]
this lgtm, but is not a regression and is only failing in the "run ten times" job... it can wait for 1.24

**Comment by cheftako** [comment]
/milestone v1.24
We can cherry pick this back to 1.23 once the PR lands.

**Pull request merged**
